src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3105,7 +3105,11 @@ internal override int CompareKeys(int index1, int index2)
                 if (next == null) return index1 - index2;
                 return next.CompareKeys(index1, index2);
             }
-            return descending ? -c : c;
+            // -c will result in a negative value for int.MaxValue (-int.MinValue == int.MinValue).
+            // Flipping keys earlier is more likely to trigger something strange in a comparer,
+            // particularly as it comes to the sort being stable.
+            // 0.CompareTo(c) is always negative for a positive c and positive for a negative.
+            return descending ? 0.CompareTo(c) : c;
```

Comment:
Would it be clearer (not needing the last line of the comment) and potentially slightly faster (though hopefully the backend compiler will do the right thing) to have:
```C#
return descending ? (c < 0 ? 1 : -1) : c;
```
?